### PR TITLE
Flow base class updates

### PR DESCRIPTION
**Pull request by EthanMarx**
1. Removes torch deps from conda, and adds them to the `train` project `pyproject.toml`
2. Removes unneeded jupyter notebook (#45)
3. Adds context dim inference #42 to flow base class
4. Adds `n_ifos` as shape parameter to architectures #43 
5. Updates `trainer.py` to accomodate the above

### CODE REVIEW

File: libs/architectures/mlpe/architectures/flows/flow.py [hunk 1]
```diff
@@ -32,10 +50,16 @@ def linear_block(self) -> transforms.Transform:
     def distribution(self) -> distributions.Distribution:
         pass
 
+    @property
+    def context_dim(self):
```
Comment (deepchatterjeeligo) [inline libs/architectures/mlpe/architectures/flows/flow.py:54]: let's keep `sample_tensor` as an optional argument with a default of what you have. Sometimes it is useful for debugging

File: libs/architectures/mlpe/architectures/flows/flow.py [hunk 2]
```diff
@@ -32,10 +50,16 @@ def linear_block(self) -> transforms.Transform:
...
+        sample_tensor = torch.zeros((1, self.n_ifos, self.strain_dim))
```
Comment (deepchatterjeeligo) [inline libs/architectures/mlpe/architectures/flows/flow.py:55]: Will this respect device change?

File: libs/trainer/mlpe/trainer/trainer.py [hunk 3]
```diff
@@ -110,15 +110,14 @@ def train(
...
+    weight_decay: float = 0.0,
+    early_stop: Optional[int] = None,
     # misc params
     device: Optional[str] = None,
     use_amp: bool = False,
     profile: bool = False,
-    optimizer_fn: Callable = torch.optim.Adam,
-    optimizer_kwargs: dict = dict(weight_decay=0),
-    scheduler_fn: Callable = torch.optim.lr_scheduler.CosineAnnealingLR,
-    scheduler_kwargs: dict = dict(eta_min=1e-5, T_max=10000),
```
Comment (deepchatterjeeligo) [inline libs/trainer/mlpe/trainer/trainer.py:121]: Why take this out? Let's just remove the lines from the project config now until the typeo trick is discovered.

File: libs/trainer/mlpe/trainer/trainer.py [hunk 4]
```diff
@@ -191,15 +190,15 @@ def train(
         f["parameters"] = parameters.cpu().numpy()
 
     param_dim = parameters.shape[-1]
-    context_dim = strain.shape[-1] * strain.shape[-2]
+    _, n_ifos, strain_dim = strain.shape
```
Comment (deepchatterjeeligo) [inline libs/trainer/mlpe/trainer/trainer.py:193]:
```suggestion
    param_dim, n_ifos, strain_dim = strain.shape
```

File: libs/trainer/mlpe/trainer/trainer.py [hunk 5]
```diff
@@ -191,15 +190,15 @@ def train(
         f["parameters"] = parameters.cpu().numpy()
 
     param_dim = parameters.shape[-1]
```
Comment (deepchatterjeeligo) [inline libs/trainer/mlpe/trainer/trainer.py:192]:
```suggestion

```

File: libs/trainer/mlpe/trainer/trainer.py [hunk 6]
```diff
@@ -222,8 +221,13 @@ def train(
     logging.info(flow)
     logging.info("Initializing loss and optimizer")
 
-    optimizer = optimizer_fn(flow.parameters(), lr=lr, **optimizer_kwargs)
```
Comment (deepchatterjeeligo) [inline libs/trainer/mlpe/trainer/trainer.py:225]: Let's not revert this bit

File: projects/sandbox/pyproject.toml [hunk 7]
```diff
@@ -72,10 +73,8 @@ batches_per_epoch = 200
 max_epochs = 300
 
 lr = 4e-3
-min_lr = 1e-5
-decay_steps = 10000
-weight_decay = 1e-5
-early_stop = 50
+
+early_stop = 100
```
Comment (deepchatterjeeligo) [inline projects/sandbox/pyproject.toml:77]: why this change?

File: libs/trainer/mlpe/trainer/trainer.py (hunk 3 above)
Comment (EthanMarx) [reply libs/trainer/mlpe/trainer/trainer.py:121]: Would unfortunately require additional `typeo` development as far as I'm aware. For now let's stick with the hard coded values(feel free to change it to exponential, or whatever you want).

File: libs/trainer/mlpe/trainer/trainer.py (hunk 4 above)
Comment (EthanMarx) [reply libs/trainer/mlpe/trainer/trainer.py:193]: No, so the parameters are separate. That dimension is the "batch" dimension (which we don't use)

File: libs/trainer/mlpe/trainer/trainer.py (hunk 5 above)
Comment (EthanMarx) [reply libs/trainer/mlpe/trainer/trainer.py:192]: See above comment

File: projects/sandbox/pyproject.toml (hunk 7 above)
Comment (EthanMarx) [reply projects/sandbox/pyproject.toml:77]: To let the network train for longer. Feel free to change. Like I said before, the parameters that are version controlled are more or less irrelevant until we find a configuration that works

File: libs/trainer/mlpe/trainer/trainer.py [hunk 8]
```diff
@@ -290,14 +294,18 @@ def train(
...
                 since_last_improvement += 1
+
                 if since_last_improvement >= early_stop:
                     logging.info(
                         "No improvement in validation loss in {} "
                         "epochs, halting training early".format(early_stop)
                     )
                     break
+            else:
+                continue
```
[the remark this answers is not in the document]
Comment (EthanMarx) [reply libs/trainer/mlpe/trainer/trainer.py:308]: Is it standard to remove the `else:` entirely, or should I replace the `continue` with `pass`?

File: libs/trainer/mlpe/trainer/trainer.py (hunk 6 above)
Comment (EthanMarx) [reply libs/trainer/mlpe/trainer/trainer.py:225]: I think we should for clarity (this is what we have in BBHnet). Given the optimizer is not our limiting factor, I think we can hold off on this until we have a better solution

File: libs/architectures/mlpe/architectures/flows/flow.py (hunk 1 above)
Comment (EthanMarx) [reply libs/architectures/mlpe/architectures/flows/flow.py:54]: Good idea

File: libs/architectures/mlpe/architectures/flows/flow.py (hunk 2 above)
Comment (EthanMarx) [reply libs/architectures/mlpe/architectures/flows/flow.py:55]: Not sure. Haven't thought much about device logic, but is definitely necessary

File: libs/architectures/mlpe/architectures/flows/flow.py (hunk 1 above)
Comment (EthanMarx) [reply libs/architectures/mlpe/architectures/flows/flow.py:56]: Actually on second thought, if someone needs to debug, they can simply instantiate the `embedding` and pass tensors through there. I think this should stay as a `property` of the instantiated flow.

File: libs/architectures/mlpe/architectures/flows/flow.py (hunk 2 above)
Comment (deepchatterjeeligo) [reply libs/architectures/mlpe/architectures/flows/flow.py:57]: well sample_tensor is on the default cpu device. If the flow is on cuda, this is expected to break.

File: libs/trainer/mlpe/trainer/trainer.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (deepchatterjeeligo) [reply libs/trainer/mlpe/trainer/trainer.py:308]: If you are thinking syntactic reason, we don't need an else if we have an elif

File: libs/trainer/mlpe/trainer/trainer.py (hunk 3 above)
Comment (deepchatterjeeligo) [reply libs/trainer/mlpe/trainer/trainer.py:121]: It does not break anything right now.

Comment (deepchatterjeeligo) [reply libs/trainer/mlpe/trainer/trainer.py:121]: You just need to not mention it in the config

File: libs/trainer/mlpe/trainer/trainer.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (EthanMarx) [reply libs/trainer/mlpe/trainer/trainer.py]: Got it. I realized my logic is wrong anyways. Fixed

File: libs/trainer/mlpe/trainer/trainer.py (hunk 6 above)
Comment (EthanMarx) [reply libs/trainer/mlpe/trainer/trainer.py:225]: Yeah, unfortunately I think we'll have to do this by hand for now given `typeo` constraints

File: projects/sandbox/pyproject.toml (hunk 7 above)
Comment (deepchatterjeeligo) [reply projects/sandbox/pyproject.toml:77]: OK. Fine with me.

File: libs/trainer/mlpe/trainer/trainer.py (hunk 3 above)
Comment (EthanMarx) [reply libs/trainer/mlpe/trainer/trainer.py:121]: Only because `typeo` is using the specified default. If you were to pass it as a config in the `pyproject.toml,` it would break. Since using the default is essentially equivalent to hard coding, lets just do that for now

File: libs/architectures/mlpe/architectures/flows/flow.py (hunk 2 above)
Comment (EthanMarx) [reply libs/architectures/mlpe/architectures/flows/flow.py:57]: Let me write some tests for this

File: libs/trainer/mlpe/trainer/trainer.py (hunk 3 above)
Comment (EthanMarx) [reply libs/trainer/mlpe/trainer/trainer.py:121]: I guess my philosophy is that the everything in the production repo should work as expected. As a hack in a branch this is of course totally fine.

**Comment by EthanMarx** [comment]
Another addition: 
Updates `ml4gw` and uses `ml4gw` `distributions` module

File: libs/trainer/mlpe/trainer/trainer.py (hunk 3 above)
Comment (deepchatterjeeligo) [reply libs/trainer/mlpe/trainer/trainer.py:121]: your philosophy is respected even without this change, and requires no breaking change in terms of typeo.

File: libs/trainer/mlpe/trainer/trainer.py (hunk 6 above)
Comment (deepchatterjeeligo) [reply libs/trainer/mlpe/trainer/trainer.py:225]: I fail to understand what the constraint is.

**Comment by deepchatterjeeligo** [comment]
Is the following the only issue BTW? That is dictionary arguments like `--scheduler-kwargs` cannot be passed like this? Because this is easily resolved by reading the keyword args as a string and running `eval`. While not general, it will work for the `scheduler_kwargs` argument in our `train` function.

```
train --background-path /home/deep.chatterjee/pe-base-dir/background.h5  \
        --waveform-dataset /home/deep.chatterjee/pe-base-dir/signals.h5 \
        --logdir /home/deep.chatterjee/pe-base-dir/log \
        --inference-params ["frequency", "quality", "hrss", "dec", "psi", "phi"] \
        --ifos ["H1", "L1"] \
        --sample-rate 2048 \
        --trigger-distance 0 \
        --kernel-length 2 \
        --fduration 1 \
        --highpass 32 \
        --batches-per-epoch 200 \
        --batch-size 512 \
        --device "cuda" \
        --max-epochs 100 \
        --init-weights /home/deep.chatterjee/pe-base-dir/training \
        --scheduler-fn "torch.optim.lr_scheduler.ExponentialLR" \
        --scheduler-kwargs "{gamma = 0.999}" \
        --verbose \
        maf \
        --num-transforms 10 \
```

fails with
```
main: error: argument --scheduler-kwargs: invalid dict value: '{gamma = 0.999}'
```

I want this flexibility since I am getting good convergence running with multiple schedules (at least until we build in the feature of a scheduler cycle which does all of this from the config). At least a shell script with multiple train lines with different schedulers with achieve this. Whatever solution gets to this point I am happy to work with.

**Comment by EthanMarx** [comment]
Yes the basic problem is that for Dictionary arguments, `typeo` can currently only take one type for the values. (e.g `Dict[str, str]`, `Dict[str, int]`, `Dict[str, float]`, but not,`Dict[str, Any]`). 

The only reason I say this is "broken" is that seeing the train function annotated with

```python
scheduler_kwargs: Dict[str, Any],
optimizer_kwargs: Dict[str, Any]
```
someone would expect to be able to specify these in the `typeo` config, but doing so would break the pipeline.

I'll change it back for the sake of getting this in and open up an issue

**Review by deepchatterjeeligo — APPROVED** [review]
LGTM.
